### PR TITLE
upgrading to LSP 5.3.0 and Monaco 0.17.0

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

- upgrading to LSP 5.3.0
  - LSP dependencies are targeting es6 now, we are forced to use babel in order to transpile es6 to es5 classes for these dependencies
- upgrading to Monaco 0.17.0
  - [x] open a CQ for all copied code to align with latest vscode.d.ts - it was already covered by previous CQs
- [x] reference peek cannot be closed with `esc`
- [x] monaco configuration does not reflect editor preferences
- [x] fix editor preview tests
- [x] depends on https://github.com/TypeFox/monaco-languageclient/pull/174
- [x] https://github.com/theia-ide/theia/pull/5901#issuecomment-522919840
- [ ] https://github.com/theia-ide/theia/pull/5901#issuecomment-528796882
- [x] https://github.com/theia-ide/theia/pull/5901#issuecomment-529311130

#### How to test

- test language features with native Theia extension:
  - test that features are available in the editor context menu and in quick command palette
  - a declaration is revealed when following the reference
  - focus belongs to a proper editor after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - reference peek is still opened after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - esc closes reference peek widget
  - completion items are properly resolved and inserted
  - focus belongs to a proper editor on peek definition (for internal and outside references, from editor to editor preview and view versa)
  - check that rename work inside the file and across files
  - check that arguments are proposed by the signature help
  - check that hover is present with properly rendered docs
  - check that symbol occurrences is highlighted when a declaration or reference is selected
  - check that go to (type|implementation) definition works properly, i.e. definition is revealed
  - check that code lens work properly
    - run/debug code lenses: https://github.com/kasecato/vscode-javadebug-sample
  - check that code actions work properly (quick fixes)
  - check that document and selection formatting work properly
  - check color provider with css colors
  - check link provider with links in html
  - check that folding ranges are properly provided for js, css and html
  - check go to symbol command
  - check open workspace symbol command
- test language features with VS Code extension:
  - typesctipt VS Code extension for testing: [ts_plugins.zip](https://github.com/theia-ide/theia/files/3513239/ts_plugins.zip)
  - to run it:
    - uninstall native typescript extension in example package.json
    - build the example app
    - unzip archive into plugins folder
    - start the example app
  - use the same checks as for native Theia extension
- test quick palette:
  - proper keys are displayed for the quick command palette, i.e. they are displayed and a corresponding Monaco command can be triggered by such keybinding
  - use sample vscode extension
  - use fuzzy search
- test vscode tree view (that context keys work properly)
- test plugin editor decorations, for example with power mode extension
- test debug console input editor
- test debug breakpoint editor
- test that preferences are propagated as config key values: change editor preferences and check whether they are reflected
- test language auto detection
- test `edit` menu with editor
- test `selection` menu with editor
- test editor context menu
- smoke test some VS Code extensions like go, python, just use editor for a bit, check backend and frontend logs for errors

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
@svenefftinge pointed out that babel won't work for plugin system :( I've opened https://github.com/theia-ide/theia/issues/5902 to try to move es6 completely instead.

Alternatively, we can move `PluginMessageReader` and `PluginMessageWriter` to `@theia/plugin-host` package and transpile it to `es6`. Later we can put all code required for host plugin process in this package to make it self-contained for deployment. cc @theia-ide/plugin-system

File: packages/plugin-ext/src/plugin/type-converters.ts [hunk 1]
```diff
@@ -455,6 +462,60 @@ export function toInternalCommand(external: theia.Command): model.Command {
         }));
 }
 
+export namespace ParameterInformation {
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/type-converters.ts]: add copyright, link to copied code and include in a CQ

File: packages/plugin-ext/src/plugin/languages/signature.ts [hunk 2]
```diff
@@ -13,32 +13,68 @@
  *
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+// copied and modified from https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L771
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/languages/signature.ts]: include in a CQ

**Comment by akosyakov** [comment]
I've tested language features with native ts extension and VS Code extension. It fixed what found.

It would be good if more VS Code extensions are tested:
- for language features (go, rust, python, ...), cc @evidolob @tolusha 
- for plugin tree views (vscode sample extension, kubernetes) - it's important that context keys are still working, cc @azatsarynnyy @tetchel 
- for editor decorations, for example power mode
- for the quick palette, for example sample extension to check that exposed monaco APIs are still there, cc @vinokurig @JPinkney 

@theia-ide/core if you have a chance please spend time to verify language features without and with vscode extensions, since we expose quite a lot of internal Monaco apis there is very high chance that it is break

I'm going to merge it only after 0.10.0, but as soon as possible that we have the entire September before 0.11.0 to catch as much as possible. cc @marcdumais-work @svenefftinge

**Comment by akosyakov** [comment]
typesctipt VS Code extension for testing: [ts_plugins.zip](https://github.com/theia-ide/theia/files/3513239/ts_plugins.zip)

To run it:
- uninstall native typescript extension in example package.json
- build the example app
- create `plugins` folder and unzip archive into it
- start the example app

**Comment by tolusha** [comment]
I checked with `go` and `python` extensions and found a couple of errors. Not sure if they related to extensions themselves. The functionality of extensions are not broken.

```
caught (in promise) TypeError: modeService.getOrCreateModeByFilenameOrFirstLine is not a function
    at MonacoEditor.<anonymous> (:3000/20.bundle.js:567)
    at step (:3000/20.bundle.js:93)
    at Object.next (:3000/20.bundle.js:74)
    at :3000/20.bundle.js:68
    at new Promise (<anonymous>)
    at push.../../packages/monaco/lib/browser/monaco-editor.js.__awaiter (:3000/20.bundle.js:64)
    at MonacoEditor.push.../../packages/monaco/lib/browser/monaco-editor.js.MonacoEditor.detectLanguage (:3000/20.bundle.js:559)
    at PluginContributionHandler.push.../../packages/plugin-ext/lib/main/browser/plugin-contribution-handler.js.PluginContributionHandler.handleContributions (:3000/89.bundle.js:8056)
    at HostedPluginSupport.<anonymous> (:3000/89.bundle.js:1541)
    at step (:3000/89.bundle.js:1390)
```

```
root ERROR [hosted-plugin: 315] With stack trace: Error: [UriError]: Scheme contains illegal characters.
    at _validateUri (http://localhost:3000/vs/editor/editor.main.js:14952:19)
    at _URI.URI (http://localhost:3000/vs/editor/editor.main.js:15045:17)
    at new _URI (http://localhost:3000/vs/editor/editor.main.js:15247:51)
    at Function.URI.revive (http://localhost:3000/vs/editor/editor.main.js:15233:30)
    at DecorationCSSRules.getCSSTextForModelDecorationGlyphMarginClassName (http://localhost:3000/vs/editor/editor.main.js:97176:96)
    at DecorationCSSRules._buildCSS (http://localhost:3000/vs/editor/editor.main.js:97079:40)
    at new DecorationCSSRules (http://localhost:3000/vs/editor/editor.main.js:97021:18)
    at createCSSRules (http://localhost:3000/vs/editor/editor.main.js:96921:29)
    at new DecorationTypeOptionsProvider (http://localhost:3000/vs/editor/editor.main.js:96944:41)
    at MonacoEditorService.CodeEditorServiceImpl.registerDecorationType (http://localhost:3000/vs/editor/editor.main.js:96853:32)
Promise rejection not handled in one second: Error: [UriError]: Scheme contains illegal characters.
```

```
navigatable.ts:93 Uncaught (in promise) TypeError: uri.withoutFragment is not a function
    at EditorManager.../../packages/core/lib/browser/navigatable.js.NavigatableWidgetOpenHandler.serializeUri (navigatable.ts:93)
    at EditorManager.../../packages/core/lib/browser/navigatable.js.NavigatableWidgetOpenHandler.createWidgetOptions (navigatable.ts:85)
    at EditorManager.../../packages/core/lib/browser/widget-open-handler.js.WidgetOpenHandler.getOrCreateWidget (widget-open-handler.ts:163)
    at EditorManager.<anonymous> (widget-open-handler.ts:81)
    at step (widget-open-handler.ts:15)
    at Object.next (widget-open-handler.ts:15)
    at widget-open-handler.ts:15
    at new Promise (<anonymous>)
    at ../../packages/core/lib/browser/widget-open-handler.js.__awaiter (widget-open-handler.ts:15)
    at EditorManager.../../packages/core/lib/browser/widget-open-handler.js.WidgetOpenHandler.open (widget-open-handler.ts:80)
```
```
root ERROR Uncaught Exception:  Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
root ERROR Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/home/tolusha/projects/theia-ide/theia/node_modules/express/lib/response.js:767:10)
    at ServerResponse.contentType (/home/tolusha/projects/theia-ide/theia/node_modules/express/lib/response.js:595:15)
    at ServerResponse.send (/home/tolusha/projects/theia-ide/theia/node_modules/express/lib/response.js:145:14)
    at /home/tolusha/projects/theia-ide/theia/packages/plugin-ext/lib/hosted/node/plugin-reader.js:112:37
    at /home/tolusha/projects/theia-ide/theia/node_modules/express/lib/response.js:430:22
    at Immediate.<anonymous> (/home/tolusha/projects/theia-ide/theia/node_modules/express/lib/response.js:1069:7)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5
```

**Comment by akosyakov** [comment]
@tolusha thank you so much 🙏 I will have a look into them

**Comment by JPinkney** [comment]
I tested a few different things:

I tried executing the following commands from both the command palette and from their keybinding and everything seemed to work:
- add cursor below/add cursor above
- delete line
- goto line
- go to symbol
- show hover
- fold all
- go to next problem
- go to previous problem

One issue is clicking enter on Git clone was not working. I was not able to git clone by going quick palette -> git clone -> https://github.com/microsoft/vscode-extension-samples.git then pressing enter

I tried testing with a sample hello world extension that calls monaco commands. While using vscode.commands.executeCommand('editor.fold'); I got no handler found but monaco.editor.fold works. I assume thats just because the prefix of monaco is still being registered though

**Comment by azatsarynnyy** [comment]
@akosyakov I've checked it with VS Code Kubernetes extension. The context menus work well.

**Comment by vinokurig** [comment]
@akosyakov 
Tested with https://github.com/microsoft/vscode-extension-samples/tree/master/quickinput-sample, the plugin works as expected

**Comment by AlexTugarev** [comment]
#### Alex' test plan

- [ ] test language features with native Theia extension:
  - [ ] test that features are available in the editor context menu and in quick command palette
  - [ ] a declaration is revealed when following the reference
  - [ ] focus belongs to a proper editor after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - [ ] reference peek is still opened after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - [ ] esc closes reference peek widget
  - [ ] completion items are properly resolved and inserted
  - [ ] focus belongs to a proper editor on peek definition (for internal and outside references, from editor to editor preview and view versa)
  - [ ] check that rename work inside the file and across files
  - [ ] check that arguments are proposed by the signature help
  - [ ] check that hover is present with properly rendered docs
  - [ ] check that symbol occurrences is highlighted when a declaration or reference is selected
  - [ ] check that go to (type|implementation) definition works properly, i.e. definition is revealed
  - [ ] check that code lens work properly
    - [ ] run/debug code lenses: https://github.com/kasecato/vscode-javadebug-sample
  - [ ] check that code actions work properly (quick fixes)
  - [ ] check that document and selection formatting work properly
  - [ ] check color provider with css colors
  - [ ] check link provider with links in html
  - [ ] check that folding ranges are properly provided for js, css and html
  - [ ] check go to symbol command
  - [ ] check open workspace symbol command
- [ ] test language features with VS Code extension:
  - [ ] typesctipt VS Code extension for testing: [ts_plugins.zip](https://github.com/theia-ide/theia/files/3513239/ts_plugins.zip)
  - [ ] to run it:
    - [ ] uninstall native typescript extension in example package.json
    - [ ] build the example app
    - [ ] unzip archive into plugins folder
    - [ ] start the example app
  - [ ] use the same checks as for native Theia extension
- [ ] test quick palette:
  - [ ] proper keys are displayed for the quick command palette, i.e. they are displayed and a corresponding Monaco command can be triggered by such keybinding
  - [ ] use sample vscode extension
  - [ ] use fuzzy search
- [ ] test vscode tree view (that context keys work properly)
- [ ] test plugin editor decorations, for example with power mode extension
- [ ] test debug console input editor
- [ ] test debug breakpoint editor

**Comment by akosyakov** [comment]
> esc closes reference peek widget FAILED

@AlexTugarev @svenefftinge how bad it is? i still cannot get why it works VS Code, but not in Monaco. In both cases they install bubbling event handlers, the only difference that in VS Code on windows, in Monaco on the editor node. But it should not be matter since the target element belongs to the editor node and should respond first. In Monaco it happens, so `esc` does not close the reference peek, but in VS Code it does not hit.

**Comment by tsmaeder** [comment]
@akosyakov we could implement a couple of API items when the new Monaco lands. How do you see the likelyhood of this PR being merged in the next week (before Sep 10)?

**Comment by akosyakov** [comment]
@tsmaeder i hope it will be merged by then, i will have a look again tomorrow what can be done with `esc` issue.

**Comment by svenefftinge** [comment]
@akosyakov as discussed yesterday, I think the [esc] issue should be solved. I'd find it very annoying and believe I'm not alone.

**Comment by akosyakov** [comment]
I found a reason: I have not synced new editor configuration with our preferences, so `editor.stablePeek` is enabled for us, but not in VS Code.

**Comment by akosyakov** [comment]
@tolusha I was able to reproduce
`caught (in promise) TypeError: modeService.getOrCreateModeByFilenameOrFirstLine is not a function` and `root ERROR Uncaught Exception:  Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client`, fixing it

Do you remember how you get `root ERROR [hosted-plugin: 315] With stack trace: Error: [UriError]: Scheme contains illegal characters.` and `navigatable.ts:93 Uncaught (in promise) TypeError: uri.withoutFragment is not a function`

**Comment by kittaakos** [comment]
> navigatable.ts:93 Uncaught (in promise) TypeError: uri.withoutFragment is not a function

@akosyakov, that issue seems to be unrelated to your monaco changes. I had that weeks ago already. See: https://github.com/TypeFox/theia-xtext-sprotty-example/issues/19

**Comment by akosyakov** [comment]
@kittaakos looks like JSON-object passed instead of proper URI, i can believe it can happen through the plugin system when we forgot to call `URI.revive`. Wonder how does it happen in sprotty case.

**Comment by tolusha** [comment]
1. Start Theia with Go [1] plugin. 
2. Open any go file in the editor.
3. Observer the error in the console:
```
root ERROR [hosted-plugin: 15960] Promise rejection not handled in one second: Error: [UriError]: Scheme contains illegal characters.
root ERROR [hosted-plugin: 15960] With stack trace: Error: [UriError]: Scheme contains illegal characters.
    at _validateUri (http://localhost:3000/vs/editor/editor.main.js:14952:19)
    at _URI.URI (http://localhost:3000/vs/editor/editor.main.js:15045:17)
    at new _URI (http://localhost:3000/vs/editor/editor.main.js:15247:51)
    at Function.URI.revive (http://localhost:3000/vs/editor/editor.main.js:15233:30)
    at DecorationCSSRules.getCSSTextForModelDecorationGlyphMarginClassName (http://localhost:3000/vs/editor/editor.main.js:97176:96)
    at DecorationCSSRules._buildCSS (http://localhost:3000/vs/editor/editor.main.js:97079:40)
    at new DecorationCSSRules (http://localhost:3000/vs/editor/editor.main.js:97021:18)
    at createCSSRules (http://localhost:3000/vs/editor/editor.main.js:96921:29)
    at new DecorationTypeOptionsProvider (http://localhost:3000/vs/editor/editor.main.js:96944:41)
    at MonacoEditorService.CodeEditorServiceImpl.registerDecorationType (http://localhost:3000/vs/editor/editor.main.js:96853:32)
```

[1] https://marketplace.visualstudio.com/items?itemName=ms-vscode.Go

**Comment by tolusha** [comment]
'TypeError: uri.withoutFragment is not a function'
I didn't do any special and I can't reproduce it anymore.

**Comment by marechal-p** [comment]
On Firefox (68.0.2 - Ubuntu 16.04) the mouse scroll is really slow (maybe 10 times slower?)

I tried capturing a gif, but I cannot scroll fast enough for it to show on the capture.

Chrome (76.0.3809.132) is not affected.

**Comment by akosyakov** [comment]
@marechal-p Could you try on Monaco playground: https://microsoft.github.io/monaco-editor/playground.html and if you reproduce file an issue for Monaco please.

**Comment by marechal-p** [comment]
Ok, turns out to be a monaco-editor bug indeed. Filed an issue.

**Comment by akosyakov** [comment]
I've managed to fix `esc` issue by registering all Monaco command in our registry. We were doing it before but only for editor actions, not for global command like `closeReferenceWidget`. Now, we handle keybindings before Monaco and call `closeReferenceWidget`.

**Comment by akosyakov** [comment]
@marechal-p I've discussed scrolling issue with @svenefftinge, decided that is annoying but not critical, you can still use mouse. Hope they fix it soon and we migrate again. We need fresh Monaco to support latest LSP api, fix memory leaks and so on.

**Comment by akosyakov** [comment]
@tolusha Do you use some particular go project for testing? if i just create go file in theia repo nothing happens.

**Comment by akosyakov** [comment]
Got another error with Go:
```
root ERROR TypeError: Cannot add property source.organizeImports, object is not extensible
    at https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:14684:38
    at Array.forEach (<anonymous>)
    at mixin (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:14672:33)
    at https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:14676:29
    at Array.forEach (<anonymous>)
    at Object.mixin (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:14672:33)
    at Configuration.CommonEditorConfiguration.updateOptions (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:81298:40)
    at StandaloneEditor.CodeEditorWidget.updateOptions (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:103757:33)
    at StandaloneEditor.updateOptions (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/vs/editor/editor.main.js:126146:44)
    at MonacoEditorProvider.updateMonacoEditorOptions (https://3000-c949f0b0-9946-4f3f-87b8-b600fb88e1c2.ws-eu0.gitpod.io/29.bundle.js:1302:33)
```

**Comment by akosyakov** [comment]
@tolusha I think i was able to address all your issues.

**Comment by tolusha** [comment]
@akosyakov 
yeap, confirm

**Comment by akosyakov** [comment]
@theia-ide/core @svenefftinge @marcdumais-work @tsmaeder Are we good to merge it? Or someone wants to test it more?

**Comment by akosyakov** [comment]
Found an error while trying auto detect language:
```
logger-protocol.ts:112 root ERROR Error: command 'monaco.editor.action.formatDocument' not found
    at StandaloneCommandService.executeCommand (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/vs/editor/editor.main.js:125588:39)
    at MonacoCommandService.<anonymous> (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:366:79)
    at step (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:299:23)
    at Object.next (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:280:53)
    at https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:274:71
    at new Promise (<anonymous>)
    at push.../../packages/monaco/lib/browser/monaco-command-service.js.__awaiter (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:270:12)
    at MonacoCommandService.push.../../packages/monaco/lib/browser/monaco-command-service.js.MonacoCommandService.executeCommand (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:357:16)
    at MonacoEditorProvider.<anonymous> (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:1328:68)
    at step (https://3000-d1c1c089-866a-49c2-a92a-7ce3a110b0cf.ws-eu0.gitpod.io/29.bundle.js:1092:23)
```

**Comment by tsmaeder** [comment]
I am all for merging this one: we'll need it to make progress on our goal to be fully VS Code API-compabible.
However, I simply don't have the capacity to do a proper code review (including testing) this week or next, so can't really approve it.
That said, I think due diligence has been done to make sure we're not seriously broken.

**Comment by svenefftinge** [comment]
What about https://github.com/theia-ide/theia/pull/5901#issuecomment-528796882 ?

**Comment by kittaakos** [comment]
Does this PR covers `CreateFile | RenameFile | DeleteFile` as a workspace edit? I do not see any support for this in the monaco workspace. Thanks!

CC: @akosyakov

**Comment by danidoedel** [comment]
@marechal-p The slow scrolling in Firefox should be fixed with Monaco 0.18.1
@akosyakov Are there any plans to upgrade in the near future?

**Pull request merged**